### PR TITLE
Add Buffer#transferTo(FileChannel, long, int)

### PR DESCRIPTION
**Pull request by yawkat**
Motivation:

Writing to a file with an explicit position allows concurrent transfers.

Modifications:

Very similar to #12396. Add #transferTo method with position, implementations are basically the same as transferTo without position. Copy the unit tests with a position arg.

Result:

Concurrent writes of different buffers to the same file using transferTo are now possible.

### CODE REVIEW

File: buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java [hunk 1]
```diff
@@ -574,6 +574,34 @@ public int transferTo(WritableByteChannel channel, int length) throws IOExceptio
...
+        if (!isAccessible()) {
+            throw bufferIsClosed(this);
+        }
+        length = Math.min(readableBytes(), length);
+        if (length == 0) {
+            return 0;
+        }
+        checkReadBounds(readerOffset(), length);
+        int totalBytesWritten = 0;
+        try (var iterator = forEachComponent()) {
```
Comment (hyperxpro) [inline buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java:588]: Let's not use `var` and go with the explicit object type.

File: buffer/src/main/java/io/netty5/buffer/Buffer.java [hunk 2]
```diff
@@ -357,6 +357,23 @@ default int writableBytes() {
...
     int transferTo(WritableByteChannel channel, int length) throws IOException;
 
+    /**
+     * Read from this buffer and write to the given channel at the given position.
+     * The number of bytes actually written to the channel are returned.
+     * No more than the given {@code length} of bytes, or the number of {@linkplain #readableBytes() readable bytes},
+     * will be written to the channel, whichever is smaller.
+     * The channel's position is not modified.
+     * The {@linkplain #readerOffset() reader-offset} of this buffer will likewise be advanced by the number of bytes
+     * written.
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty5/buffer/Buffer.java:367]:
```suggestion
     * The {@linkplain #readerOffset() reader-offset} of this buffer will, however, be advanced by the number of bytes
     * written.
```

File: buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java [hunk 3]
```diff
@@ -574,6 +574,34 @@ public int transferTo(WritableByteChannel channel, int length) throws IOExceptio
...
+            ByteBuffer[] byteBuffers = new ByteBuffer[countReadableComponents()];
+            int counter = 0;
+            for (var c = iterator.firstReadable(); c != null; c = c.nextReadable()) {
+                byteBuffers[counter++] = c.readableBuffer();
+            }
+            int bufferCount = countAndPrepareBuffersForChannelIO(length, byteBuffers);
+            for (int i = 0; i < bufferCount; i++) {
+                int bytesWritten = channel.write(byteBuffers[i], addExact(position, totalBytesWritten));
+                totalBytesWritten = addExact(totalBytesWritten, bytesWritten);
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java:597]: This looks suspicious, both here and in the code you copied. We should either exit the loop and return the total, or avoid advancing `i`, if `byteBuffers[I].remaining() != 0`.

File: buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java (hunk 1 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java:588]: Yeah, this is fine.

**Review by chrisvest — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty5/buffer/Buffer.java (hunk 2 above)
Comment (yawkat) [reply buffer/src/main/java/io/netty5/buffer/Buffer.java:367]: i'll change it manually and in the old transferTo too

File: buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java (hunk 3 above)
Comment (yawkat) [reply buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java:597]: good point, i will try to make a test

Comment (yawkat) [reply buffer/src/main/java/io/netty5/buffer/DefaultCompositeBuffer.java:602]: fixed and added a test, though it's a bit ugly for FileChannel.

File: buffer/src/main/java/io/netty5/buffer/Buffer.java (hunk 2 above)
Comment (yawkat) [reply buffer/src/main/java/io/netty5/buffer/Buffer.java]: ahh i see.

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
@yawkat The `MemSegBuffer` implementation also needs to be updated, that's why the Java 20 build failed:

```
2023-07-17T17:32:57.6427969Z [INFO] -------------------------------------------------------------
2023-07-17T17:32:57.6428662Z [ERROR] COMPILATION ERROR : 
2023-07-17T17:32:57.6429425Z [INFO] -------------------------------------------------------------
2023-07-17T17:32:57.6433393Z [ERROR] /code/buffer-memory-segment/src/main/java/io/netty5/buffer/memseg/MemSegBuffer.java:[49] error: MemSegBuffer is not abstract and does not override abstract method transferTo(FileChannel,long,int) in Buffer
2023-07-17T17:32:57.6434298Z [INFO] 1 error
```

**Comment by chrisvest** [comment]
@yawkat The `buffer-memory-segment` module has a test-jar dependency on the `buffer` module, so all the buffer tests run again in the `buffer-memory-segment` module.

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
Thanks!

**Pull request merged**